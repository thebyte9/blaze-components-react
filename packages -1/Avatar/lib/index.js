"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var utils_1 = __importDefault(require("@blaze-react/utils"));
var react_1 = __importStar(require("react"));
var Avatar = utils_1["default"](function (_a) {
    var _b;
    var modifier = _a.modifier, url = _a.url, username = _a.username, classNames = _a.utils.classNames, attr = __rest(_a, ["modifier", "url", "username", "utils"]);
    var _c = react_1.useState(url), avatarUrl = _c[0], setAvatar = _c[1];
    var _d = react_1.useState(false), validUrl = _d[0], setValidUrl = _d[1];
    var avatarClassName = classNames("avatar", (_b = {},
        _b["avatar--" + modifier] = !!modifier,
        _b));
    var initials = !username
        ? "!"
        : username
            .split(" ")
            .map(function (subName) { return subName[0].toUpperCase(); })
            .join("")
            .substring(0, 2);
    if (url && !validUrl) {
        var imageData = new Image();
        imageData.src = url;
        imageData.addEventListener("load", function () {
            setAvatar(url);
            setValidUrl(true);
        });
    }
    return (react_1["default"].createElement("div", { className: avatarClassName }, validUrl ? (react_1["default"].createElement("img", __assign({ src: avatarUrl, alt: "avatar" }, attr))) : (react_1["default"].createElement("span", null, initials))));
});
var availableModifiers = {
    med: "med",
    small: "small",
    xSmall: "x-small"
};
Avatar.availableModifiers = availableModifiers;
Avatar.defaultProps = {
    modifier: "",
    url: "",
    username: ""
};
exports["default"] = Avatar;
