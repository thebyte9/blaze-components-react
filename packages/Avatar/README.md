## Description

Avatar is the graphical representation of the user or the user's alter ego or character.

## Usage

- Image

```js
const url = 'http://lorempixel.com/400/400/people/';

<Avatar url={url} modifier={Avatar.availableModifiers.med} />

<Avatar
  url={url}
  username="Ismael Haytam"
  modifier={Avatar.availableModifiers.small}
/>
```

- User Initials

```js
<Avatar username="Blaz 2" modifier={Avatar.availableModifiers.med} />

<Avatar
  username="Lorem Ipsum"
  modifier={Avatar.availableModifiers.xSmall}
/>
```

## API

##### Avatar can receive a number of `props` as follow:

| NAME     |  TYPE  | DEFAULT | Options             |
| :------- | :----: | :-----: | ------------------- |
| modifier | String |  empty  | med, small, x-small |
| url      | string |  empty  |
| username | String |    !    |
