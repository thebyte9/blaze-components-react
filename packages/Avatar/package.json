{
    "name": "@blaze-react/avatar",
    "version": "0.0.1",
    "description": "",
    "main": "index.js",
    "scripts": {
        "build": "./node_modules/.bin/babel index.js -d lib",
        "build-storybook": "build-storybook -c .storybook -o docs-build",
         "lint": "eslint ./ --ignore-path ../../.eslintignore",

        "test": "jest",
        "test:coverage": "jest --coverage",
        "pre-publish": "npm run build && npm run lint && npm run test"
    },
    "keywords": [],
    "author": "",
    "license": "ISC",
    "dependencies": {},
    "devDependencies": {
        "@babel/cli": "^7.4.4",
        "@babel/core": "^7.4.5"
    },
    "peerDependencies": {
        "react": "^16.8.6",
        "react-dom": "^16.8.6",
        "prop-types": "^15.7.2"
    },
    "babel": {
        "presets": [
            "@babel/preset-env",
            "@babel/preset-react"
        ]
    }
}