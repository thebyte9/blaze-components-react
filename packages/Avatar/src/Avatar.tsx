import buildClassNames from '../../Utils/src/buildClassNames';
import React, { useState } from "react";

type Tmodfier = "med" | "small" | "x-small";

interface IAvatarProps {
  modifier?: Tmodfier;
  url?: string;
  username?: string;
  attr?: Record<string, unknown>;
}

const Avatar =
  ({
    modifier,
    url,
    username,
    ...attr
  }: IAvatarProps): JSX.Element => {
    const [avatarUrl, setAvatar] = useState<string | undefined>(url);
    const [validUrl, setValidUrl] = useState<boolean>(false);

    const avatarClassName: string = buildClassNames("avatar", {
      [`avatar--${modifier}`]: !!modifier
    });

    const initials = !username
      ? "!"
      : username
        .split(" ")
        .map(subName => subName[0].toUpperCase())
        .join("")
        .substring(0, 2);

    if (url && !validUrl) {
      const imageData = new Image();
      imageData.src = url;

      imageData.addEventListener(
        "load",
        (): void => {
          setAvatar(url);
          setValidUrl(true);
        }
      );
    }

    return (
      <div className={avatarClassName}>
        {validUrl ? (
          <img src={avatarUrl} alt="avatar" {...attr} />
        ) : (
          <span>{initials}</span>
        )}
      </div>
    );
  }
  ;

const availableModifiers = {
  med: "med",
  small: "small",
  xSmall: "x-small"
};

Avatar.availableModifiers = availableModifiers;

Avatar.defaultProps = {
  modifier: "",
  url: "",
  username: ""
};

export default Avatar;
