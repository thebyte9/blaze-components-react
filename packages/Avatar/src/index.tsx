import React, { FunctionComponent, useState } from "react";
interface IAvatarProps {
  modifier?: string;
  url?: string;
  username?: string;
  attrs?: any;
}
const Avatar: FunctionComponent<IAvatarProps> = ({
  modifier,
  url,
  username,
  ...attr
}): JSX.Element => {
  const [avatarUrl, setAvatar] = useState<string | undefined>(url);
  const [invalidUrl, setInvalidUrl] = useState<boolean>(true);
  const buildedModifier = modifier ? `avatar--${modifier}` : "";


  const getInitilas = (): string => {
    if (!username) {
      return '!';
    }
    return username
      .split(" ")
      .map(subName => subName[0].toUpperCase())
      .join("")
      .substring(0, 2);
  }

  const initials: string = getInitilas();

  if (url && invalidUrl) {
    const imageData = new Image();
    imageData.src = url;

    imageData.addEventListener('load', (): void => {
      setAvatar(url)
      setInvalidUrl(false);
    });
  }

  return (
    <div className={`avatar ${buildedModifier}`}>
      {avatarUrl ? (
        <img src={avatarUrl} alt="avatar" {...attr} />
      ) : (
          <span>{initials}</span>
        )}
    </div>
  );
};

Avatar.defaultProps = {
  modifier: "",
  url: "",
  username: ""
};
export default Avatar;
