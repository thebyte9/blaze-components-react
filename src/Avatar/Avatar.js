import React, {
  Fragment, useState, useRef, useEffect
} from 'react';
import PropTypes from 'prop-types';

const Avatar = ({
  modifier,
  url,
  username,
  ...attr
}) => {
  const [size, setSize] = useState(0);
  const [avatarUrl, setAvatar] = useState(url);

  const _modifier = modifier ? `avatar--${modifier}` : '';
  const ref = useRef(null);

  useEffect(() => setSize(ref.current.clientHeight / 2));

  const initials = username && (username.split(' ')
    .map(subName => subName[0]
      .toUpperCase()).join(''))
    .substring(0, 2);
  
  const imageData = new Image();
  imageData.src = url;
  imageData.onload = () => setAvatar(url);

  return (
    <Fragment>
      <div className={`avatar ${_modifier}`} ref={ref}>
        {avatarUrl && <img src={avatarUrl} alt="avatar" className="avatar__icon" {...attr} />}
        <div className="avatar__image">
          {avatarUrl && <img src={avatarUrl} alt="alt text here" {...attr} />}
        </div>
        {!avatarUrl && <span style={{ fontSize: `${size}px` }}>{initials}</span>}
      </div>
    </Fragment>
  );
};

Avatar.propTypes = {
  modifier: PropTypes.string,
  url: PropTypes.string,
  username: PropTypes.string
};

Avatar.defaultProps = {
  modifier: '',
  url: '',
  username: '!'
};

export default Avatar;
