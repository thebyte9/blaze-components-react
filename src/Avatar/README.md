
## Description

Avatar is the graphical representation of the user or the user's alter ego or character.

## Usage

* Image

```js
const url = 'http://lorempixel.com/400/400/people/';

<Avatar url={url} modifier="med"/>
<Avatar url={url} username="Ismael Haytam" modifier="small"/>
```

* User Initials

``` js
<Avatar username="Blaz 2"/>
<Avatar username="Kongan Page" modifier="x-small"/>
```

## API

##### Avatar can receive a number of `props` as follow:


| NAME   | TYPE | DEFAULT | Description |
| :---  | :---:  | :---: | ------- |
| url | string | '' |   Image Url |
| modifier | String | '' | To control image size: med, small, x-small |
| username | String | ! | If the image does not load, the initials of the username will be shown |

