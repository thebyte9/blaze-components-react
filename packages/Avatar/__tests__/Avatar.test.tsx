import { mount } from 'enzyme';
import expect from 'expect';
import React from 'react';
import Avatar from '../src';

declare let window: any;

const url = 'http://lorempixel.com/400/400/people/';

const defaultProps = (override: object = {}) => ({
  modifier: 'med',
  url,
  username: 'Blaze 2',
  ...override
});

describe('Avatar component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const wrapper = mount(<Avatar {...defaultProps()} />);
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  // test('should load image', () => {

  //   const dummyFileReader = {
  //     addEventListener: jest.fn((_, evtHandler) => { evtHandler(); }),
  //   };

  //   window.Image = jest.fn(() => dummyFileReader);

  //   mount(<Avatar {...defaultProps()} />);

  // });

  test('should be defined and renders also when the image does not load', () => {
    const wrapper = mount(<Avatar url={`notFound/${url}`} modifier="med" />);
    expect(wrapper).toBeDefined();
  });

  test('The text of the avatar must be the initials of the username', () => {
    const overrid = {
      modifier: '',
      url: ''
    };

    const wrapper = mount(<Avatar {...defaultProps(overrid)} />);
    const span = wrapper.find('span');
    expect(span.text()).toContain('B2');
  });
});
