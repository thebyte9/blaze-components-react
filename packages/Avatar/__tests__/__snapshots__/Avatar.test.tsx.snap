// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component should be defined and renders correctly (snapshot) 1`] = `
<withUtils()
  modifier="med"
  url="http://lorempixel.com/400/400/people/"
  username=""
>
  <Component
    modifier="med"
    url="http://lorempixel.com/400/400/people/"
    username=""
    utils={
      Object {
        "ErrorMessage": [Function],
        "classNames": [Function],
        "removeExtraSpaces": [Function],
        "uniqueId": [Function],
      }
    }
  >
    <div
      className="avatar avatar--med"
    >
      <img
        alt="avatar"
        src="http://lorempixel.com/400/400/people/"
      />
    </div>
  </Component>
</withUtils()>
`;
